core/lib/merkle_tree/src/domain.rs

```diff
@@ -61,6 +61,12 @@ impl ZkSyncTree {
             .expect("failed initializing `rayon` thread pool")
     }
 
+    /// Returns tree pruner and a handle to stop it
+    pub fn pruner(&self) -> (MerkleTreePruner<RocksDBWrapper>, MerkleTreePrunerHandle) {
```

Comment:
- Bikeshedding: Please move this method below the constructor methods (i.e., `new_with_mode()`), so that getters are logically grouped.
- Don't quite like this interface because it easily allows creating multiple pruners for the same tree. It would be more idiomatic to integrate creating pruner into the tree constructor (probably via a builder since otherwise the constructor would become too messy). OTOH, it's certainly not the highest-priority issue; it's possible to clone the underlying `RocksDBWrapper` and misuse the tree APIs in this way.

Use of a hypothetical builder:

```rust
let builder = ZkSyncTree::builder(db);
let (pruner, pruner_handle) = builder.enable_pruning(); // takes &self
let tree = builder.build(); // or .build_lightweight() for the lightweight tree
```

...Then, you would stash `(pruner, pruner_handle)` in `AsyncTree` and allow to take them from it.

Reply:
I don't like the use of builder pattern here, for me it looks awkward that ZkSyncTree builder can build different type of object. And how it's currently done is consistent with reader constructor. If we really want to make sure there is only one pruner, imo it makes more sense to leave the interface like it is, I've added a guardrail though